File: hf2-mf-uhf/python/example_pid_advisor_pll.py
# Copyright 2017 Zurich Instruments AG

"""
Zurich Instruments LabOne Python API Example

Demonstrate how to connect to a Zurich Instruments Lock-in Amplifier and use
the PID Advisor to set up an internal PLL control loop using ziDAQServer's
pidAdvisor Module.
Connect to a Zurich Instruments Lock-in Amplifier and obtain optimized
P, I, and D parameters for an internal PLL loop using ziDAQServer's pidAdvisor module.

Requirements:
    * LabOne Version >= 20.02
    * Instruments:
        1 x UHF or an MF with the PID Option
    * signal output 1 connected to signal input 1 with a BNC cable.

Usage:
    example_pid_advisor_pll.py [options] <device_id>
    example_pid_advisor_pll.py -h | --help

Arguments:
    <device_id>  The ID of the device [device_type: UHF.*(PID)|MF.*(PID)]

Options:
    -h --help              Show this screen.
    -s --server_host IP    Hostname or IP address of the dataserver [default: localhost]
    -p --server_port PORT  Port number of the data server [default: 8004]
    --no-plot              Hide plot of the recorded data.

Raises:
    Exception     If the specified devices do not match the requirements.
    RuntimeError  If the devices is not "discoverable" from the API.

See the LabOne Programming Manual for further help:
https://docs.zhinst.com/labone_programming_manual/
"""

import time
import numpy as np
import zhinst.utils
import matplotlib.pyplot as plt


def run_example(
    device_id: str,
    server_host: str = "localhost",
    server_port: int = 8004,
    plot: bool = True,
):
    """run the example."""

    apilevel_example = 6  # The API level supported by this example.
    (daq, device, props) = zhinst.utils.create_api_session(
        device_id, apilevel_example, server_host=server_host, server_port=server_port
    )
    zhinst.utils.api_server_version_check(daq)

    # This example additionally requires two oscillators.
    if props["devicetype"].startswith("MF") and ("MD" not in props["options"]):
        raise RuntimeError(
            "Required option set not satisfied. "
            "On MF Instruments this example requires both the PID and the MD "
            f"Option. Device `{device}` reports devtype `{props['devicetype']}` \
                and options `{props['options']}`."
        )

    # Create a base configuration: Disable all available outputs, awgs, demods, scopes,...
    zhinst.utils.disable_everything(daq, device)

    # PID configuration.
    target_bw = 10e3  # Target bandwidth (Hz).
    pid_index = 0  # PID index.
    pid_input = 3  # PID input (3 = Demod phase).
    pid_input_channel = 0  # Demodulator number.
    setpoint = 0  # Phase setpoint.
    phase_unwrap = True
    pid_output = 2  # PID output (2 = oscillator frequency).
    pid_output_channel = 0  # The index of the oscillator controlled by PID.
    pid_center_frequency = 400e5  # (Hz).
    pid_limits = 100e5  # (Hz).

    # Define configuration and configure the device.
    out_channel = 0
    osc_index = 1
    # Get the value of the instrument's default Signal Output mixer channel.
    out_mixer_channel = zhinst.utils.default_output_mixer_channel(props)
    amplitude = 1.0

    # now the settings relevant to this experiment
    exp_setting = [
        ["/%s/oscs/%d/freq" % (device, osc_index), pid_center_frequency],
        ["/%s/sigouts/%d/on" % (device, out_channel), True],
        ["/%s/sigouts/%d/enables/%d" % (device, out_channel, out_mixer_channel), True],
        ["/%s/sigouts/%d/range" % (device, out_channel), 1.0],
        [
            "/%s/sigouts/%d/amplitudes/%d" % (device, out_channel, out_mixer_channel),
            amplitude,
        ],
        ["/%s/pids/%d/input" % (device, pid_index), pid_input],
        ["/%s/pids/%d/inputchannel" % (device, pid_index), pid_input_channel],
        ["/%s/pids/%d/setpoint" % (device, pid_index), setpoint],
        ["/%s/pids/%d/output" % (device, pid_index), pid_output],
        ["/%s/pids/%d/outputchannel" % (device, pid_index), pid_output_channel],
        ["/%s/pids/%d/center" % (device, pid_index), pid_center_frequency],
        ["/%s/pids/%d/enable" % (device, pid_index), False],
        ["/%s/pids/%d/phaseunwrap" % (device, pid_index), phase_unwrap],
        ["/%s/pids/%d/limitlower" % (device, pid_index), -pid_limits],
        ["/%s/pids/%d/limitupper" % (device, pid_index), pid_limits],
    ]
    daq.set(exp_setting)

    # Perform a global synchronisation between the device and the data server:
    # Ensure that the settings have taken effect on the device before starting the pid_advisor.
    daq.sync()

    # set up PID Advisor
    pid_advisor = daq.pidAdvisor()

    pid_advisor.set("device", device)
    # Turn off auto-calc on param change. Enabled
    # auto calculation can be used to automatically
    # update response data based on user input.
    pid_advisor.set("auto", False)
    pid_advisor.set("pid/targetbw", target_bw)

    # PID advising mode (bit coded)
    # bit 0: optimize/tune P
    # bit 1: optimize/tune I
    # bit 2: optimize/tune D
    # Example: mode = 7: Optimize/tune PID
    pid_advisor.set("pid/mode", 7)

    # PID index to use (first PID of device: 0)
    pid_advisor.set("index", pid_index)

    # DUT model
    # source = 1: Lowpass first order
    # source = 2: Lowpass second order
    # source = 3: Resonator frequency
    # source = 4: Internal PLL
    # source = 5: VCO
    # source = 6: Resonator amplitude
    dut_source = 4
    pid_advisor.set("dut/source", dut_source)

    # IO Delay of the feedback system describing the earliest response
    # for a step change. This parameter does not affect the shape of
    # the DUT transfer function
    pid_advisor.set("dut/delay", 0.0)

    # Other DUT parameters (not required for the internal PLL model)
    # pid_advisor.set('dut/gain', 1.0)
    # pid_advisor.set('dut/bw', 1000)
    # pid_advisor.set('dut/fcenter', 15e6)
    # pid_advisor.set('dut/damping', 0.1)
    # pid_advisor.set('dut/q', 10e3)

    # Start values for the PID optimization. Zero
    # values will imitate a guess. Other values can be
    # used as hints for the optimization process.
    pid_advisor.set("pid/p", 0)
    pid_advisor.set("pid/i", 0)
    pid_advisor.set("pid/d", 0)

    # Start the module thread
    pid_advisor.execute()

    # Advise
    pid_advisor.set("calculate", 1)
    print("Starting advising. Optimization process may run up to a minute...")
    calculate = 1

    t_start = time.time()
    t_timeout = t_start + 90
    while calculate == 1:
        time.sleep(0.1)
        calculate = pid_advisor.getInt("calculate")
        progress = pid_advisor.progress()
        print(f"Advisor progress: {progress[0]:.2%}.", end="\r")
        if time.time() > t_timeout:
            pid_advisor.finish()
            raise Exception("PID advising failed due to timeout.")
    print("")
    print(f"Advice took {time.time() - t_start:0.1f} s.")

    # Get all calculated parameters.
    result = pid_advisor.get("*", True)
    # Check that the dictionary returned by poll contains the data that are needed.
    assert result, "pidAdvisor returned an empty data dictionary?"

    if result is not None:
        # Now copy the values from the PID Advisor to the device's PID.
        pid_advisor.set("todevice", 1)
        # Let's have a look at the optimised gain parameters.
        p_advisor = result["/pid/p"][0]
        i_advisor = result["/pid/i"][0]
        d_advisor = result["/pid/d"][0]
        print(
            f"The pidAdvisor calculated the following gains, \
                P: {p_advisor}, I: {i_advisor}, D: {d_advisor}."
        )

    if plot:

        plt.close("all")
        bode_complex_data = result["/bode"][0]["x"] + 1j * result["/bode"][0]["y"]
        bode_grid = result["/bode"][0]["grid"]

        step_x = result["/step"][0]["x"]
        step_grid = result["/step"][0]["grid"]

        bw_advisor = result["/bw"][0]

        _, axes = plt.subplots(2, 1)
        axes[0].plot(bode_grid, 20 * np.log10(np.abs(bode_complex_data)))
        axes[0].set_xscale("log")
        axes[0].grid(True)
        axes[0].set_title(
            r"Model response for internal PLL with "
            "P = %0.1f, I = %0.1f,\n"
            "D = %0.5f and bandwidth %0.1f kHz"
            % (p_advisor, i_advisor, d_advisor, bw_advisor * 1e-3)
        )
        axes[0].set_ylabel("Bode Gain (dB)")
        axes[0].autoscale(enable=True, axis="x", tight=True)

        axes[1].plot(bode_grid, np.angle(bode_complex_data) / np.pi * 180)
        axes[1].set_xscale("log")
        axes[1].grid(True)
        axes[1].set_xlabel("Frequency (Hz)")
        axes[1].set_ylabel("Bode Phase (deg)")
        axes[1].autoscale(enable=True, axis="x", tight=True)

        _, axis = plt.subplots(1, 1)
        axis.plot(step_grid * 1e6, step_x)
        axis.grid(True)
        axis.set_title(
            r"Step response for internal PLL with "
            "P = %0.1f, I = %0.1f,\nD = %0.5f and bandwidth %0.1f kHz"
            % (p_advisor, i_advisor, d_advisor, bw_advisor * 1e-3)
        )
        axis.set_xlabel(r"Time ($\mu$s)")
        axis.set_ylabel(r"Step Response")
        axis.autoscale(enable=True, axis="x", tight=True)
        axis.set_ylim([0.0, 1.05])
        plt.draw()
        plt.show()


if __name__ == "__main__":
    import sys
    from pathlib import Path

    cli_util_path = Path(__file__).resolve().parent / "../../utils/python"
    sys.path.insert(0, str(cli_util_path))
    cli_utils = __import__("cli_utils")
    cli_utils.run_commandline(run_example, __doc__)
    sys.path.remove(str(cli_util_path))

File: hf2-mf-uhf/matlab/example_pid_advisor_pll.m
function result = example_pid_advisor_pll(device_id, varargin)
% EXAMPLE_PID_ADVISOR Setup a PID for internal PLL mode
%
% USAGE SAMPLE = EXAMPLE_PID_ADVISOR(DEVICE_ID)
%
% Setup the PID for the internal PLL mode on the device specified by
% DEVICE_ID. DEVICE_ID should be a string, e.g. 'dev1000' or 'uhf-dev1000'.
%
% NOTE This example can only be ran on MF or UHF Instruments with the PID
% option enabled.
%
% NOTE Additional configuration: Connect signal output 1 to signal input 1
% with a BNC cable.
%
% NOTE Please ensure that the ziDAQ folders 'Driver' and 'Utils' are in your
% Matlab path. To do this (temporarily) for one Matlab session please navigate
% to the ziDAQ base folder containing the 'Driver', 'Examples' and 'Utils'
% subfolders and run the Matlab function ziAddPath().
% >>> ziAddPath;
%
% Use either of the commands:
% >>> help ziDAQ
% >>> doc ziDAQ
% in the Matlab command window to obtain help on all available ziDAQ commands.
%
% Copyright 2008-2018 Zurich Instruments AG

clear ziDAQ;

if ~exist('device_id', 'var')
  error(['No value for device_id specified. The first argument to the ' ...
    'example should be the device ID on which to run the example, ' ...
    'e.g. ''dev1000'' or ''uhf-dev1000''.'])
end

% Check the ziDAQ MEX (DLL) and Utility functions can be found in Matlab's path.
if ~(exist('ziDAQ') == 3) && ~(exist('ziCreateAPISession', 'file') == 2)
  fprintf('Failed to either find the ziDAQ mex file or ziDevices() utility.\n')
  fprintf('Please configure your path using the ziDAQ function ziAddPath().\n')
  fprintf('This can be found in the API subfolder of your LabOne installation.\n');
  fprintf('On Windows this is typically:\n');
  fprintf('C:\\Program Files\\Zurich Instruments\\LabOne\\API\\MATLAB2012\\\n');
  return
end

% The API level supported by this example.
apilevel_example = 6;
% Create an API session; connect to the correct Data Server for the device.
required_err_msg = ['This example only runs with MF or UHF Instruments with ' ...
                    'the PID Option enabled.'];
[device, props] = ziCreateAPISession(device_id, apilevel_example, ...
                                     'required_devtype', 'UHFLI|(MF.*)', ...
                                     'required_options', {'PID'}, ...
                                     'required_err_msg', required_err_msg);
ziApiServerVersionCheck();

%% Define some sensible parameters based on the device type.
if strfind(props.devicetype, 'MF')
% The target bandwidth [Hz].
  pid_center_frequency = 450e3;
  default_target_bw = 10e3;
  pid_limits = 100e3; % [Hz]
elseif strfind(props.devicetype, 'UHF')
  pid_center_frequency = 10.0e6;
  default_target_bw = 100e3;
  pid_limits = 1e6; % [Hz]
else
  error(['Unexpected device type `' props.devicetype '`.']);
end

%% Define parameters relevant to this example. Default values specified by the
% inputParser below are overwritten if specified as name-value pairs via the
% `varargin` input argument.
p = inputParser;
isnonnegscalar = @(x) isnumeric(x) && isscalar(x) && (x > 0);
p.addParamValue('target_bw', default_target_bw, isnonnegscalar);
p.parse(varargin{:});

% Create a base configuration: Disable all available outputs, awgs,
% demods, scopes,...
ziDisableEverything(device);

%% Configure the PID as a PLL
% Select phase as PID input for the internal PLL mode
ziDAQ('setInt', ['/' device '/pids/0/input'], 3)
ziDAQ('setInt', ['/' device '/pids/0/inputchannel'], 3)
% Set the phase setpoint to 0deg for the internal PLL mode
ziDAQ('setDouble', ['/' device '/pids/0/setpoint'], 0)
% Enable phase unwrap mode for internal PLL mode
ziDAQ('setInt', ['/' device '/pids/0/phaseunwrap'], 1)
% Select oscillator as output of the PID for the internal PLL mode
ziDAQ('setInt', ['/' device '/pids/0/output'], 2)
ziDAQ('setInt', ['/' device '/pids/0/outputchannel'], 0)
% Set the center frequency and limits
ziDAQ('setDouble', ['/' device '/pids/0/center'],  pid_center_frequency)
ziDAQ('setDouble', ['/' device '/pids/0/limitlower'], -pid_limits)
ziDAQ('setDouble', ['/' device '/pids/0/limitupper'],  pid_limits)

% Start of the PID advisor module
advisor = ziDAQ('pidAdvisor');

%% Setup of PID advisor
ziDAQ('set', advisor, 'device', device)
% Turn off auto-calc on param change. Enabled
% auto calculation can be used to automatically
% update response data based on user input.
ziDAQ('set', advisor, 'auto', 0);
ziDAQ('set', advisor, 'pid/targetbw', p.Results.target_bw)
% PID advising mode (bit coded)
% bit 0: optimize/tune P
% bit 1: optimize/tune I
% bit 2: optimize/tune D
% Example: mode = 3: Only optimize/tune PI
ziDAQ('set', advisor, 'pid/mode', 3)
% PID index to use (first PID of device: 0)
ziDAQ('set', advisor, 'index', 0);
% DUT model
% source = 1: Lowpass first order
% source = 2: Lowpass second order
% source = 3: Resonator frequency
% source = 4: Internal PLL
% source = 5: VCO
% source = 6: Resonator amplitude
dut_source = 4;
ziDAQ('set', advisor, 'dut/source', 4)
% IO Delay of the feedback system describing the earliest response
% for a step change. This parameter does not affect the shape of
% the DUT transfer function
ziDAQ('set', advisor, 'dut/delay', 0.0)
% Start values for the PID optimization. Zero
% values will initate a guess. Other values can be
% used as hints for the optimization process.
% Following parameters are not required for the internal PLL model
% ziDAQ('set', advisor, 'dut/gain', 1)
% ziDAQ('set', advisor, 'dut/bw', 1000)
% ziDAQ('set', advisor, 'dut/fcenter', 15e6)
% ziDAQ('set', advisor, 'dut/damping', 0.1)
% ziDAQ('set', advisor, 'dut/q', 10e3)
ziDAQ('set', advisor, 'pid/p', 0);
ziDAQ('set', advisor, 'pid/i', 0);
ziDAQ('set', advisor, 'pid/d', 0);
ziDAQ('set', advisor, 'calculate', 0)

% Start the module thread
ziDAQ('execute', advisor);

%% Advise
fprintf('Starting advising. Optimization process may run up to a minute...\n');
ziDAQ('set', advisor, 'calculate', 1)
calculate = 1;
timeout = 60; % [s]
tic;
while calculate == 1
  pause(0.5)
  calculate = ziDAQ('getInt', advisor, 'calculate');
  if toc > timeout
    ziDAQ('finish', advisor);
    ziDAQ('clear', advisor);
    error('PID advising failed due to timeout.')
  end
end
fprintf('Advice took %0.1fs\n', toc);


%% Get all calculated parameters
result = ziDAQ('get', advisor, '*');

if ~isempty(result)
  %% Transfer the PID coefficients to the PID
  % Now copy the values from the PID Advisor to the device's PID.
  ziDAQ('set', advisor, 'todevice', 1)
  fprintf('The pidAdvisor calculated the following gains,  P: %f, I: %f, D: %f.\n', ...
          result.pid.i, result.pid.d, result.pid.d);
  % You may now also want to enable the device's PID:
  % ziDAQ('setInt', ['/' device '/pids/0/enable'], 1)

  %% Plot calculated PID response
  complexData = result.bode.x + 1i * result.bode.y;
  subplot(3, 1, 1)
  h = semilogx(result.bode.grid, 20 * log10(abs(complexData)));
  set(h, 'Color', 'black')
  set(h, 'LineWidth', 1.5)
  box on
  grid on
  xlabel('Frequency [Hz]')
  ylabel('Bode Gain [dB]')
  title(sprintf('Calculated model response for internal PLL with P = %0.0f, I = %0.0f, D = %0.5f and bandwidth %0.0fkHz\n', result.pid.p, result.pid.i, result.pid.d, result.bw * 1e-3))
  subplot(3, 1, 2)
  h = semilogx(result.bode.grid, angle(complexData) / pi * 180);
  set(h, 'Color', 'black')
  set(h, 'LineWidth', 1.5)
  box on
  grid on
  xlabel('Frequency [Hz]')
  ylabel('Bode Phase [deg]')
  subplot(3, 1, 3)
  h = plot(result.step.grid * 1e6, result.step.x);
  set(h, 'Color', 'black')
  set(h, 'LineWidth', 1.5)
  box on
  grid on
  xlabel('Time [us]')
  ylabel('Step Response')
end

% Release module resources. Especially important if modules are created
% inside a loop to prevent excessive resource consumption.
ziDAQ('clear', advisor);

end
